File: docs/_code/producer.cpp
#include <diaspora/Driver.hpp>
#include <diaspora/TopicHandle.hpp>
#include <iostream>

int main(int argc, char** argv) {

    if(argc != 3) {
        std::cerr << "Usage: "
                  << argv[0] << " <groupfile> <topic>" << std::endl;
        return -1;
    }

    auto group_file = argv[1];
    auto topic_name = argv[2];

    try {

        diaspora::Metadata options;
        options.json()["group_file"] = group_file;
        options.json()["margo"] = nlohmann::json::object();
        options.json()["margo"]["use_progress_thread"] = true;

        diaspora::Driver driver = diaspora::Driver::New("mofka", options);

        diaspora::TopicHandle topic = driver.openTopic(topic_name);

        diaspora::Producer producer = topic.producer();
        for(size_t i = 0; i < 100; ++i) {
            auto future = producer.push(
                diaspora::Metadata{R"({"x":42,"name":"bob"})"},
                diaspora::DataView{}
            );
            // auto event_id = future.wait();
        }
        producer.flush();

    } catch(const diaspora::Exception& ex) {
        std::cerr << ex.what() << std::endl;
    }
    return 0;
}

File: docs/_code/producer.py
import sys
import pymargo.core
from pymargo.core import Engine
from diaspora_stream.api import Driver


def produce(group_file: str, topic_name: str):
    options = {
        "group_file": group_file,
        "margo": {
            "use_progress_thread": True
        }
    }
    driver = Driver.new("mofka", options)
    topic = driver.open_topic(topic_name)
    producer = topic.producer()

    for i in range(0, 100):
        future = producer.push(
            metadata={"x": i*42, "name": "bob"},
            data=bytes())
        # event_id = future.wait()

    producer.flush()


if __name__ == "__main__":

    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <groupfile> <topic>")
    group_file = sys.argv[1]
    topic_name = sys.argv[2]
    produce(group_file, topic_name)
